drivers/gpio/gpio_pca95xx.c

```diff
@@ -0,0 +1,675 @@
+ * @param value Value to set (0 or 1)
+ *
+ * @return 0 if successful, failed otherwise
+ */
+static int gpio_pca95xx_write(struct device *dev, int access_op,
+				u32_t pin, u32_t value)
+{
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->driver_data;
+	union gpio_pca95xx_port_data *reg_out = &drv_data->reg_cache.output;
+	int ret;
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	/* Invert input value for pins configurated as active low. */
+	switch (access_op) {
+	case GPIO_ACCESS_BY_PIN:
+		if (value) {
+			reg_out->all |= BIT(pin);
+		} else {
+			reg_out->all &= ~BIT(pin);
+		}
+		break;
+	case GPIO_ACCESS_BY_PORT:
+		reg_out->all = value;
+		break;
+	default:
+		ret = -ENOTSUP;
+		goto done;
+	}
+
+	ret = write_port_regs(dev, REG_OUTPUT_PORT0, reg_out);
+
+done:
+	k_sem_give(&drv_data->lock);
+	return ret;
+}
+
+/**
+ * @brief Read the pin or port status
+ *
+ * @param dev Device struct of the PCA95XX
+ * @param access_op Access operation (pin or port)
+ * @param pin The pin number
+ * @param value Value of input pin(s)
+ *
+ * @return 0 if successful, failed otherwise
+ */
+static int gpio_pca95xx_read(struct device *dev, int access_op,
+			       u32_t pin, u32_t *value)
+{
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->driver_data;
+	union gpio_pca95xx_port_data buf;
+	int ret;
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	ret = read_port_regs(dev, REG_INPUT_PORT0, &buf);
+	if (ret != 0) {
+		goto done;
+	}
+
+	switch (access_op) {
+	case GPIO_ACCESS_BY_PIN:
+		*value = (buf.all >> pin) & 0x01;
+		break;
+	case GPIO_ACCESS_BY_PORT:
+		*value = buf.all;
+		break;
+	default:
+		ret = -ENOTSUP;
+		break;
+	}
+
+done:
+	k_sem_give(&drv_data->lock);
+	return ret;
+}
+
+static int gpio_pca95xx_port_get_raw(struct device *dev, u32_t *value)
+{
+	union gpio_pca95xx_port_data buf;
+	int ret;
+
+	ret = read_port_regs(dev, REG_INPUT_PORT0, &buf);
+	if (ret != 0) {
+		goto done;
+	}
+
+	*value = buf.all;
+
+done:
+	return ret;
+}
+
+static int gpio_pca95xx_port_set_masked_raw(struct device *dev,
+					      u32_t mask, u32_t value)
+{
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->driver_data;
+	union gpio_pca95xx_port_data *reg_out = &drv_data->reg_cache.output;
+	u32_t new_value;
+
+	new_value = (reg_out->all & ~mask) | (mask & value);
+
+	return write_out_regs_locked(dev, new_value);
+}
+
+static int gpio_pca95xx_port_set_bits_raw(struct device *dev, u32_t mask)
+{
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->driver_data;
+	union gpio_pca95xx_port_data *reg_out = &drv_data->reg_cache.output;
+	u32_t new_value;
+
+	new_value = reg_out->all | mask;
```

Comment:
If the thread gets preempted the moment we read out `reg_out` value its value can be changed by another thread. Subsequently, if this thread updates the value of `reg_out`, which happens inside `write_out_regs_locked`, the value placed there by another thread will be overwritten. We have a race condition. The full read, modify, write operation needs to be protected by the mutex. A good example of how this can be implemented can be found in `gpio_sx1509b.c` driver.